pkg/virt-operator/resource/apply/instancetypes.go

```diff
@@ -0,0 +1,134 @@
+		}
+		log.Log.V(2).Infof("instancetype %v created", instancetype.GetName())
+		return nil
+	}
+
+	if equality.Semantic.DeepEqual(foundObj.Annotations, instancetype.Annotations) &&
+		equality.Semantic.DeepEqual(foundObj.Labels, instancetype.Labels) &&
+		equality.Semantic.DeepEqual(foundObj.Spec, instancetype.Spec) {
+		log.Log.V(4).Infof("instancetype %v is up-to-date", instancetype.GetName())
+		return nil
+	}
+
+	instancetype.ResourceVersion = foundObj.ResourceVersion
+	if _, err := instancetypeClient.Update(context.Background(), instancetype, metav1.UpdateOptions{}); err != nil {
+		return fmt.Errorf("unable to update instancetype %+v: %v", instancetype, err)
+	}
+	log.Log.V(2).Infof("instancetype %v updated", instancetype.GetName())
+
+	return nil
+}
+
+func (r *Reconciler) deleteInstancetypes() error {
+	ls := labels.Set{
+		v1.AppComponentLabel: v1.AppComponent,
+		v1.ManagedByLabel:    v1.ManagedByLabelOperatorValue,
+	}
+
+	if err := r.clientset.VirtualMachineClusterInstancetype().DeleteCollection(context.Background(), metav1.DeleteOptions{}, metav1.ListOptions{
+		LabelSelector: ls.String(),
+	}); err != nil {
+		return fmt.Errorf("unable to delete preferences: %v", err)
+	}
+
+	return nil
+}
+
+func (r *Reconciler) createOrUpdatePreferences() error {
+	for _, preference := range r.targetStrategy.Preferences() {
+		if err := r.createOrUpdatePreference(preference.DeepCopy()); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *Reconciler) createOrUpdatePreference(preference *instancetypev1beta1.VirtualMachineClusterPreference) error {
+	preferenceClient := r.clientset.VirtualMachineClusterPreference()
+
+	foundObj, err := preferenceClient.Get(context.Background(), preference.Name, metav1.GetOptions{})
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+
+	imageTag, imageRegistry, id := getTargetVersionRegistryID(r.kv)
+	injectOperatorMetadata(r.kv, &preference.ObjectMeta, imageTag, imageRegistry, id, true)
+
+	if errors.IsNotFound(err) {
+		if _, err := preferenceClient.Create(context.Background(), preference, metav1.CreateOptions{}); err != nil {
+			return fmt.Errorf("unable to create preference %+v: %v", preference, err)
+		}
+		log.Log.V(2).Infof("preference %v created", preference.GetName())
+		return nil
+	}
+
+	if equality.Semantic.DeepEqual(foundObj.Annotations, preference.Annotations) &&
+		equality.Semantic.DeepEqual(foundObj.Labels, preference.Labels) &&
+		equality.Semantic.DeepEqual(foundObj.Spec, preference.Spec) {
+		log.Log.V(4).Infof("preference %v is up-to-date", preference.GetName())
+		return nil
+	}
+
+	preference.ResourceVersion = foundObj.ResourceVersion
+	if _, err := preferenceClient.Update(context.Background(), preference, metav1.UpdateOptions{}); err != nil {
+		return fmt.Errorf("unable to update preference %+v: %v", preference, err)
+	}
+	log.Log.V(2).Infof("preference %v updated", preference.GetName())
+
+	return nil
+}
```

Comment:
Hey, I managed to make your code bigger, much harder to read, and probably broken!
https://github.com/jean-edouard/kubevirt/commit/49e516e3c7de45db65f7e38b3773f5851f08c25d
More seriously, I wish there was a good way to refactor these 2 identical functions... Maybe there's a better way I don't know about?

Reply:
Generics but it's tricky and could be a follow up post `v1.1.0` before we move the feature to `beta`  in `v1.2.0` and on by default. FWIW we used some generics in the SSP operator that initially deployed these resources downstream in OKD/OCP, this functionality will be deprecated and disabled by default in the near future :

https://github.com/kubevirt/ssp-operator/blob/main/internal/operands/common-instancetypes/reconcile.go

Reply:
@jean-edouard as above, would this be something we could refine in follow ups during `v1.2.0`?